server/src/main/java/org/uiautomation/ios/wkrdp/WebKitNotificationListener.java

```diff
@@ -27,162 +35,278 @@
+      handle(applicationSentListingMessage);
+    } else if (message instanceof ApplicationDataMessage) {
+      // ApplicationDataMessage m = (ApplicationDataMessage)message;
+    } else if (message instanceof ApplicationConnectedMessage) {
+      ApplicationConnectedMessage applicationConnectedMessage = ApplicationConnectedMessage.class.cast(message);
+      handle(applicationConnectedMessage);
     }
+  }
+
+  private void handle(ReportSetupMessage reportSetupMessage) {
+    driver.setDevice(reportSetupMessage.getDevice());
+    sync.signalSimRegistered();
+  }
 
-    if (message instanceof ReportConnectedApplicationsMessage) {
-      ReportConnectedApplicationsMessage m = (ReportConnectedApplicationsMessage) message;
-      if (m.getApplications().size() == 0) {
-        log.warning("ReportConnectedApplicationsMessage reported 0 app.");
+  private void handle(ReportConnectedApplicationsMessage reportConnectedApplicationsMessage) {
+    try {
+      applicationRegistrationLock.lock();
+      registeredApplications.addAll(reportConnectedApplicationsMessage.getApplications());
+      if (!isDriverSuccessfullyNotified()) {
+        log.log(Level.WARNING,
+            "No connectable applications found in ReportConnectedApplicationsMessage, will wait before signalling driver");
+      }
+    } finally {
+      applicationRegistrationLock.unlock();
+    }
+  }
+
+  private void handle(ApplicationSentListingMessage applicationSentListingMessage) {
+    if (!isValidForProcessing(applicationSentListingMessage)) {
+      return;
+    }
+    try {
+      pagesProcessLock.lock();
+      boolean pagesSimilar = arePagesNotChanged(applicationSentListingMessage);
+      if (log.isLoggable(Level.FINE)) {
+        log.fine("pages " + (pagesSimilar ? "are equal " : "have changed ") + ": " + driver.getPages() + "-> "
+            + applicationSentListingMessage.getPages() + ": " + applicationSentListingMessage);
+      }
+      if (pagesSimilar) {
+        driver.setPages(applicationSentListingMessage.getPages()); // Update the titles.
+        return;
+      }
+      if (session == null || session.getApplication().isSafari()) {
+        processPagesForSafariApp(applicationSentListingMessage);
       } else {
-        driver.setApplications(m.getApplications());
+        processPagesForHybridApp(applicationSentListingMessage);
+      }
+    } finally {
+      pagesProcessLock.unlock();
+    }
+  }
+
+  private void handle(ApplicationConnectedMessage applicationConnectedMessage) {
+    try {
+      applicationRegistrationLock.lock();
+      registeredApplications.add(applicationConnectedMessage.getApplication());
+      if (!isDriverSuccessfullyNotified()) {
+        log.log(Level.WARNING,
+            "No connectable applications found in ApplicationConnectedMessage, will wait before signalling driver");
+      }
+    } finally {
+      applicationRegistrationLock.unlock();
+    }
+  }
+
+  private boolean isDriverSuccessfullyNotified() {
```

Comment:
@mach6 
Yes I get your point, but this whole WebKitNotificationListener is stateless. Since this method is called from two different method that handle different message types, the logging in the caller will give more information of which message type was processed during the call. I can still move this logging into isDriverSuccessfullyNotified() method, but the logging message will be plain as "no connectable application received.." or so. I am thinking that would not help lot in debugging? I am okay with either any thoughts?